### ISSUE TITLE
Proposal: Open-loop evaluation pipeline (validation split + trajectory metrics)

### ISSUE THREAD

**Issue by riita10069** [autowarefoundation/auto_e2e#66]
## Background

The training loop is now functional (#14) and the model learns on real data — trajectory imitation loss decreases reliably on both L2D and NVIDIA PhysicalAI datasets. The next step is measuring *what* the model has learned against established metrics so we can compare with published baselines and track progress.

AutoE2E outputs `(acceleration, curvature) × 64 timesteps` at 10 Hz (6.4s planning horizon). This proposal covers how to evaluate these predictions using the metrics standard in the E2E autonomous driving literature.

## 1. Trajectory Integration: (a, κ) → (x, y)

Since evaluation metrics operate in displacement space (meters), we need to integrate the model's action-space outputs into position trajectories. The standard approach is forward Euler integration with a kinematic bicycle model:

```python
x, y, theta, v = 0.0, 0.0, 0.0, v_initial  # ego-centric frame
positions = []
for t in range(T):
    x += v * cos(theta) * dt
    y += v * sin(theta) * dt
    theta += v * kappa[t] * dt  # yaw_rate = v * kappa
    v += a[t] * dt
    positions.append((x, y))
```

Where `v_initial` is the ego's speed at the current frame (available from `egomotion_history[-4]` for L2D, or the egomotion state for nvidia). `dt = 0.1s` (10 Hz).

This is the same integration used by [VADv2 (2024)](https://arxiv.org/abs/2402.13243) and other action-space planners. The gap between waypoint-based and action-space models in benchmarks is explicitly discussed in ["Addressing the Waypoint-Action Gap" (2025)](https://arxiv.org/abs/2602.06214).

## 2. Primary Metrics (Open-Loop)

Following the nuScenes planning benchmark established by [ST-P3 (ECCV 2022)](https://arxiv.org/abs/2207.07601) and adopted by [UniAD (CVPR 2023)](https://arxiv.org/abs/2212.10156), [VAD (ICLR 2024)](https://arxiv.org/abs/2303.12077), and [SparseDrive (2024)](https://arxiv.org/abs/2405.19620):

### L2 Displacement Error (ADE)

```
L2_t = || (x̂_t, ŷ_t) - (x_t, y_t) ||₂
ADE@Hs = (1/Hs) × Σ_{t=1}^{Hs} L2_t
```

Reported at horizons: **1s, 2s, 3s, 6.4s** (extending the standard 3s to our full 64-step horizon).

### Final Displacement Error (FDE)

```
FDE@Hs = L2_{Hs}
```

The L2 error at the final timestep of each horizon.

### Collision Rate

At each future timestep, place the ego's oriented bounding box at the predicted pose. Check for geometric intersection with ground-truth positions of other agents (if available). Reported as percentage of scenarios with at least one collision, per horizon.

**Note:** L2D does not provide other-agent annotations, so collision rate may only be feasible on datasets with 3D object labels (nuScenes, NVIDIA with future label releases). For L2D, we can compute a simplified proxy: off-road rate using the BEV map channel.

## 3. Comfort Metrics (Direct from Outputs)

These can be computed directly from the model's (a, κ) predictions without integration, making them especially natural for AutoE2E:

| Metric | Formula | [nuPlan threshold](https://github.com/motional/nuplan-devkit) |
|--------|---------|-----------------|
| Longitudinal jerk | `j_lon[t] = (a[t+1] - a[t]) / dt` | ≤ 4.13 m/s³ |
| Lateral acceleration | `a_lat[t] = v[t]² × κ[t]` | ≤ 4.89 m/s² |
| Lateral jerk | `j_lat[t] = (a_lat[t+1] - a_lat[t]) / dt` | ≤ 8.37 m/s³ |
| Max absolute yaw rate | `ω[t] = v[t] × κ[t]` | ≤ 0.95 rad/s |

Comfort violation rate: fraction of predictions exceeding any threshold.

### L2D

L2D does not define an official train/val split (all 100k episodes are in a single "train" partition). We need to define our own. The dataset provides rich metadata for stratified splitting:

- 30 German cities (GPS coordinates available)
- 3+ years of temporal coverage (Nov 2022 – Jan 2025)
- Driver policy labels (`EXPERT` vs student) in `task.policy`
- Road type, weather, lighting metadata per episode

**Proposed approach:** Geographic holdout — reserve 3–5 cities as validation. This ensures the model generalizes to unseen road layouts, not just unseen frames from familiar locations. The geographic/temporal leakage problem identified by [Lilja et al. (CVPR 2024)](https://arxiv.org/abs/2312.10296) in nuScenes motivates this stricter separation.

Alternatively, a simpler episode-range split (last 10% of episode indices) is acceptable for early experiments before geographic metadata is parsed.

### NVIDIA PhysicalAI

The dataset spans 25 countries and 2500+ cities. The `clip_index` and `data_collection` metadata (accessible via SDK) contain geographic groupings suitable for holdout splits. To be investigated once more clips are downloaded.

## 5. Baselines

To contextualize model performance, we need simple baselines that reveal whether the perception pipeline is contributing beyond ego-status extrapolation.

| Baseline | Description | Expected behavior |
|----------|-------------|-------------------|
| **Constant velocity** | a=0, κ=0 for all 64 steps | Lower bound. Works on highways, fails on curves/intersections. |
| **Last-observed repeat** | Repeat current (a, κ) for all steps | Slightly better on smooth roads, catastrophic on transitions. |
| **Ego-status MLP** | MLP on (past speed, accel, curvature) → future (a, κ). No perception input. | [AD-MLP (CVPR 2024)](https://arxiv.org/abs/2305.10430) showed this beats many perception-based models on L2 error. AutoE2E must clearly beat this to justify its architecture. |

The ego-status MLP is particularly important: [Zhai et al. (CVPR 2024)](https://arxiv.org/abs/2305.10430) demonstrated that ego-status alone achieves ~0.44m average L2 on nuScenes, matching or exceeding some perception-based approaches. If AutoE2E doesn't beat this baseline on L2, it means the camera inputs aren't being utilized effectively.

## 6. Reference Numbers (nuScenes Planning Benchmark)

For context, published results at 1s/2s/3s horizons:

| Method | L2-1s (m) | L2-2s (m) | L2-3s (m) | Avg | Col. Rate |
|--------|-----------|-----------|-----------|-----|-----------|
| [ST-P3 (ECCV 2022)](https://arxiv.org/abs/2207.07601) | 1.33 | 2.11 | 2.90 | 2.11 | 0.71% |
| [UniAD (CVPR 2023)](https://arxiv.org/abs/2212.10156) | 0.29 | 0.89 | 1.53 | 0.90 | 0.31% |
| [VAD (ICLR 2024)](https://arxiv.org/abs/2303.12077) | 0.17 | 0.34 | 0.60 | 0.37 | — |
| [SparseDrive (2024)](https://arxiv.org/abs/2405.19620) | 0.15 | 0.29 | 1.30 | 0.58 | 0.06% |

Note: These are on nuScenes val with 6 cameras at 2Hz waypoint output. AutoE2E uses 7 cameras at 10Hz in action space, so direct comparison requires matching the evaluation protocol (downsample predictions to 2Hz, evaluate first 3s only).

## 7. Future Direction: Closed-Loop Proxy

Open-loop L2 error has known limitations (ego inertia dominates, no error compounding). [NAVSIM (NeurIPS 2024)](https://arxiv.org/abs/2406.15349) provides a lightweight closed-loop proxy:

- Plan is executed via LQR controller in a 4s non-reactive simulation
- PDMS score: `NC × DAC × (5×TTC + 5×EP + 2×C) / 12`
- Better correlation with real closed-loop performance than raw L2 error

L2D has also announced plans for real-vehicle closed-loop evaluation (summer 2025). Once available, this would be the most meaningful evaluation for AutoE2E.

## Implementation Plan

- [ ] `Model/evaluation/metrics.py` — ADE, FDE, comfort metrics, (a, κ) → (x, y) integration utility
- [ ] `Model/evaluation/baselines.py` — constant-velocity, last-observed, ego-status MLP
- [ ] `Model/evaluation/evaluate.py` — script: load checkpoint → run val set → report metrics table
- [ ] Validation split definition (episode list or geographic holdout)
- [ ] Integration into `train.py` (periodic val evaluation during training)

## References

- [ST-P3](https://arxiv.org/abs/2207.07601) — Hu et al., "ST-P3: End-to-end Vision-based Autonomous Driving via Spatial-Temporal Feature Learning", ECCV 2022
- [UniAD](https://arxiv.org/abs/2212.10156) — Hu et al., "Planning-oriented Autonomous Driving", CVPR 2023
- [VAD](https://arxiv.org/abs/2303.12077) — Jiang et al., "VAD: Vectorized Scene Representation for Efficient Autonomous Driving", ICLR 2024
- [VADv2](https://arxiv.org/abs/2402.13243) — Chen et al., "VADv2: End-to-End Vectorized Autonomous Driving via Probabilistic Planning", 2024
- [SparseDrive](https://arxiv.org/abs/2405.19620) — Sun et al., "SparseDrive: End-to-End Autonomous Driving with Sparse Scene Representation", 2024
- [AD-MLP / Rethinking](https://arxiv.org/abs/2305.10430) — Zhai et al., "Rethinking the Open-Loop Evaluation of End-to-End Autonomous Driving in nuScenes", CVPR 2024
- [NAVSIM](https://arxiv.org/abs/2406.15349) — Dauner et al., "NAVSIM: Data-Driven Non-Reactive Autonomous Vehicle Simulation", NeurIPS 2024
- [Localization Is All You Evaluate](https://arxiv.org/abs/2312.10296) — Lilja et al., CVPR 2024
- [Waypoint-Action Gap](https://arxiv.org/abs/2602.06214) — 2025
- [nuPlan devkit](https://github.com/motional/nuplan-devkit) — comfort thresholds and closed-loop metrics
